buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/DistributionTest.kt

```diff
@@ -63,6 +66,37 @@ open class DistributionTest : Test() {
         jvmArgumentProviders.add(libsRepository)
         systemProperty("java9Home", project.findProperty("java9Home") ?: System.getProperty("java9Home"))
     }
+
+    override fun executeTests() {
+        val timer = setTimeoutMonitor()
+        super.executeTests()
+        timer?.cancel()
+    }
+
+    private
+    fun setTimeoutMonitor(): Timer? {
```

Comment:
How about this which is maybe more idiomatic Kotlin:

```kotlin
    override fun executeTests() {
        printStackAfterTimeout { super.executeTests() }
    }

    private
    fun printStackAfterTimeout(work: () -> Unit) = if (BuildEnvironment.isCiServer) {
        val timer = Timer(true).apply {
            schedule(timerTask {
                project.javaexec {
                    classpath = this@IntegrationTest.classpath
                    main = "org.gradle.integtests.fixtures.timeout.JavaProcessStackTracesMonitor"
                }
            }, determineTimeoutMillis())
        }
        try {
            work()
        } finally {
            timer.cancel()
        }
    } else {
        work()
    }
```

Reply:
Brilliant!